src/lib/coda_base/transaction_logic.ml

```diff
@@ -1516,11 +1512,35 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
         ~modify_timing:(fun acc ->
           update_timing_when_no_deduction ~txn_global_slot acc )
     in
+    let compute_balance account_id =
+      match get_user_account_with_location t account_id with
+      | Ok (`Existing _, account) ->
+          Some account.balance
+      | _ ->
+          None
+    in
+    let balances =
+      match Fee_transfer.to_singles transfer with
+      | `One ft ->
+          { User_command_status.Balance_data.fee_payer_balance=
+              compute_balance (Fee_transfer.Single.receiver ft)
+          ; source_balance= None
+          ; receiver_balance= None }
+      | `Two (ft1, ft2) ->
+          { User_command_status.Balance_data.fee_payer_balance=
+              compute_balance (Fee_transfer.Single.receiver ft1)
+          ; source_balance= None
+          ; receiver_balance=
+              compute_balance (Fee_transfer.Single.receiver ft2) }
+    in
     Undo.Fee_transfer_undo.
-      {fee_transfer= transfer; previous_empty_accounts; receiver_timing}
+      { fee_transfer= transfer
+      ; previous_empty_accounts
+      ; receiver_timing
+      ; balances }
```

Comment:
could you move the balance field out of `Undo.t`? I think of Undo.t as the state before the transaction was applied. Like have another type that has the balances and the undo as separate fields?

Reply:
Pushing back on this. This matches the behaviour of user commands, where we record their status (and implicitly balances) as part of their undo. `Transaction_applied.t` may be a more appropriate name (and `undo : Transaction_applied.t -> ...` still makes sense); I propose to do this rename in a follow-up PR.